Speaker Feedback: Add base plugin file and comment type functions

coreymckrill opened the pull request:
Gotta start somewhere.

This scaffolds the main plugin file and creates CRUD functions for managing comments of our custom type.

There isn't a function for the "D" in "CRUD" because `wp_delete_comment` is sufficient, and needs no wrapper method.

Fixes #338 

**To test**

1. Activate the plugin on a site in your local dev environment
1. Use `wp shell` to try running the CRUD functions:
  * `\WordCamp\SpeakerFeedback\Comment\get_feedback()` (do this one first to see that there aren't any submissions yet, then again after adding a submission)
  * `\WordCamp\SpeakerFeedback\Comment\add_feedback( {post ID}, array( 'name' => 'Foo', 'email' => 'foo@example.org' ), array( 'bar' => 'baz' ) )`
  * `\WordCamp\SpeakerFeedback\Comment\update_feedback( {comment ID}, array( 'bar' => 'bop', 'fizz' => 'bang' ) )`
  * `\WordCamp\SpeakerFeedback\Comment\delete_feedback( {comment ID} )`
1. To check that comment meta is getting added/updated, try accessing them as properties on the returned comment objects:
  * `$test = \WordCamp\SpeakerFeedback\Comment\get_feedback()`
  * `$test[0]->bar`

coreymckrill committed:
Add main plugin file

coreymckrill committed:
Add CRUD functions for our custom comment type

coreymckrill committed:
Add note about update function return value

coreymckrill committed:
Add sorting to comment query

coreymckrill committed:
Switch $feedback_author value to integer

Since `wp_insert_comment` takes a user ID as an argument, it's better
to just start with that, rather than a WP_User object. Otherwise we
might end up converting the user ID into an object just to feed it into
the function, where it would get converted back to an ID :/

coreymckrill committed:
Fill in doc block param details

ryelle commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/includes/comment.php:
```diff
@@ -0,0 +1,93 @@
+	// This makes loading meta values for comments much faster.
+	wp_queue_comments_for_comment_meta_lazyload( $feedback );
+
+	return $feedback;
```
Can this also return the comment meta? That'll prevent an extra step for us later.

coreymckrill replied:
I thought about that, but it seemed really inefficient to have to call `get_comment_meta` on every result in the comment array. And then I wasn't sure if I could attach the meta directly to the results, since the `WP_Comment` class is `final`. But I didn't actually test it... I'll do some more research on this.

coreymckrill replied:
Ok, taking a slightly different approach in 711e063d

I've updated the testing instructions to match.

ryelle commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/includes/comment.php:
```diff
@@ -0,0 +1,93 @@
+		'comment_approved' => 0, // "hold".
+		'comment_post_ID'  => $post_id,
+		'comment_type'     => COMMENT_TYPE,
+		'comment_meta'     => $feedback_meta,
```
Maybe when we figure out what the meta we need is, we can add an abstraction here to sanitize/format & prefix the meta keys & values… (nothing to do now, just thinking out loud)

coreymckrill replied:
Yeah, I was thinking we'd do a separate file with all the meta-related functionality, but it might make sense here to run the data through a normalizing/sanitizing function once we have that available.

ryelle reviewed (commented):
Works as expected, and these "comments" already show up in wp-admin (we'll probably want to fix that for regular comment screens, maybe as part of #340 )

> There isn't a function for the "D" in "CRUD" because wp_delete_comment is sufficient, and needs no wrapper method.

For consistency I would add it in anyway, but I don't feel too strongly about that.

coreymckrill committed:
Add delete_feedback() function to complete CRUD ops

coreymckrill commented:
> For consistency I would add [the D] in anyway

Fair enough: 9274c7d6

coreymckrill committed:
Removed unused use

coreymckrill committed:
Add wrapper class to enable extra functionality for comment objects

WP_Comment objects don't have any methods for retrieving comment meta,
and the class is defined as `final` so it can't be extended to include
additional methods. Instead, we can import each WP_Comment instance into
another object that has magic methods that allow usage of WP_Comment's
properties and methods, while also enabling retrieval of comment meta.

ryelle reviewed (approved):
Looks good, all methods work as expected 👍

coreymckrill committed:
Make namespacing more precise in comment type string
